server/src/main/java/org/elasticsearch/index/mapper/vectors/VectorSimilarityFloatValueSource.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.index.mapper.vectors;
+
+import org.apache.lucene.index.FloatVectorValues;
+import org.apache.lucene.index.LeafReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.VectorSimilarityFunction;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.DoubleValues;
+import org.apache.lucene.search.DoubleValuesSource;
+import org.apache.lucene.search.IndexSearcher;
+import org.elasticsearch.search.profile.query.QueryProfiler;
+import org.elasticsearch.search.vectors.QueryProfilerProvider;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Objects;
+
+/**
+ * DoubleValuesSource that is used to calculate scores according to a similarity function for a KnnFloatVectorField, using the
+ * original vector values stored in the index
+ */
+public class VectorSimilarityFloatValueSource extends DoubleValuesSource implements QueryProfilerProvider {
+
+    private final String field;
+    private final float[] target;
+    private final VectorSimilarityFunction vectorSimilarityFunction;
+    private long vectorOpsCount;
+
+    public VectorSimilarityFloatValueSource(String field, float[] target, VectorSimilarityFunction vectorSimilarityFunction) {
+        this.field = field;
+        this.target = target;
+        this.vectorSimilarityFunction = vectorSimilarityFunction;
+    }
+
+    @Override
+    public DoubleValues getValues(LeafReaderContext ctx, DoubleValues scores) throws IOException {
+        final LeafReader reader = ctx.reader();
+
+        FloatVectorValues vectorValues = reader.getFloatVectorValues(field);
+
+        return new DoubleValues() {
+            @Override
+            public double doubleValue() throws IOException {
+                vectorOpsCount++;
+                return vectorSimilarityFunction.compare(target, vectorValues.vectorValue());
```

Comment:
@benwtrent , FYI - this is different than the [original PR](https://github.com/elastic/elasticsearch/pull/116663/files#diff-59ec43923aaf5660d792b39a5c4d8b46209b8151db729cb833f53e6b293c3c62R50) as Lucene 10 has a different interface for FloatVectorValues vector retrieval. 

I believe we're covered by testing and the checks I've done, but flagging for additional 👀 on it.